### ISSUE TITLE
DynamicProperty stops observing dependent properties in 5.0.2

### ISSUE THREAD

**Issue by Igor-Palaguta** [ReactiveCocoa/ReactiveCocoa#3443]
Updated to RAC 5.0.2, and it stops observing dependent properties immediately.

```availableEarnings``` and ```promoBalance``` works as before, but ```availableBalance``` does not work

Can't imagine why it happens, so attaching you as much code as possible

```
@objc
protocol Wallet {
   var promoBalance: NSDecimalNumber { get }
   var availableEarnings: NSDecimalNumber { get }
   var availableBalance: NSDecimalNumber { get }
}

class ManagedWallet: NSManagedObject, Wallet {
   @NSManaged var availableEarnings: NSDecimalNumber
   @NSManaged var promoBalance: NSDecimalNumber

   dynamic var availableBalance: NSDecimalNumber {
      return self.availableEarnings.adding(self.promoBalance)
   }

   class func keyPathsForValuesAffectingAvailableBalance() -> NSSet {
      return NSSet(array: ["availableEarnings", "promoBalance"])
   }
}

struct WalletViewModel {
   private let wallet: Wallet

   let promoBalance: Property<String?>
   let availableEarnings: Property<String?>
   let availableBalance: Property<String?>

   init(wallet: Wallet) {
   
      self.wallet = wallet

      let formatMoney: (NSDecimalNumber?) -> String? = /**/

      let walletObject = wallet as! NSObject
      self.promoBalance = DynamicProperty<NSDecimalNumber>(object: walletObject, keyPath: #keyPath(Wallet.promoBalance))
         .map(formatMoney)

      self.availableEarnings = DynamicProperty<NSDecimalNumber>(object: walletObject, keyPath: #keyPath(Wallet.availableEarnings))
         .map(formatMoney)

      self.availableBalance = DynamicProperty<NSDecimalNumber>(object: walletObject, keyPath: #keyPath(Wallet.availableBalance))
         .map(formatMoney)
   }
}
```

**Comment by andersio** [comment]
This seems to be an issue of Objective-C interoperability. Try to use `NSObject.keyPathsForValuesAffectingValue(forKey:)` instead.

**Comment by Igor-Palaguta** [comment]
@andersio unfortunately same effect. Why did it work in 5.0.1? Any chance it can be fixed?

**Comment by andersio** [comment]
Have you upgraded to Swift 3.1? That could be a factor.

Make sure you are overriding the `keyPathsForValuesAffectingValue(forKey:)` class method, which returns `Set<String>`.

**Comment by andersio** [comment]
Just tried Xcode 8.3.1. This is working:

```
    class func keyPathsForValuesAffectingValuePlusOne() -> Set<String> {
        return Set([#keyPath(Test.value)])
    }
```

**Comment by Igor-Palaguta** [comment]
Still using Xcode 8.3, will try. Thanks for hint

**Comment by Igor-Palaguta** [comment]
I think now I know why it happens.

Observing works fine. But after this: https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3418, when value is deallocated RAC sends to observer nil.
```availableBalance``` returns temporary NSDecimalNumber, and shouldObserveDeinit returns true fot it. So when temporary object is deallocated nil is sent. It did not happen before. It is unpredictable, and not sure that it is correct

@andersio please share your vision

Thanks!

**Comment by andersio** [comment]
Oh. This hasn't come to my mind. It is fixed in #3435.

**Comment by andersio** [comment]
It was indeed a bug, and it was fixed in #3439 (merged). Sorry for any inconvenience caused.

**Comment by Igor-Palaguta** [comment]
Thanks it works! Waiting for RAC 5.0.3
